src/server/wesnothd/server.cpp

```diff
@@ -576,15 +576,22 @@ bool server::ip_exceeds_connection_limit(const std::string& ip) const
 	return connections >= concurrent_connections_;
 }
 
-std::string server::is_ip_banned(const std::string& ip)
+utils::optional<server_base::login_ban_info> server::is_ip_banned(const std::string& ip)
 {
-	if(!tor_ip_list_.empty()) {
-		if(find(tor_ip_list_.begin(), tor_ip_list_.end(), ip) != tor_ip_list_.end()) {
-			return "TOR IP";
-		}
+	if(!tor_ip_list_.empty() && utils::contains(tor_ip_list_, ip)) {
+		return login_ban_info{ MP_AUTH_SERVER_IP_BAN_ERROR, "TOR IP", {} };
+	}
+
+	if(auto server_ban_info = ban_manager_.get_ban_info(ip)) {
+		return login_ban_info{
+			MP_AUTH_SERVER_IP_BAN_ERROR,
+			server_ban_info->get_reason(),
+			server_ban_info->get_remaining_ban_time()
+		};
 	}
 
-	return ban_manager_.is_ip_banned(ip);
+	// TODO: try and handle forum bans here as well...
```

Comment:
I figured since the actual error reporting is basically the same, it would make sense. But it looks like the forum ban handling requires additional info that's not present at this point in the login sequence (namely, the username).